### PR TITLE
router: inherit upstream connection filter state in upstream request stream info

### LINKED ISSUE #10457: Upstream FilterState for HTTP requests
Currently the `upstreamFilterState` is only set by the TCP proxy. We should support this for HTTP as well: this would allow upstream filters to set information on the connection, which could then be consumed in HTTP access logs. 

As part of this we would probably want to make `FilterState` less downstream specific (e.g. rename the scopes to be direction agnostic).

### PR DESCRIPTION
**Pull request by snowp**
This adds the upstream connection filter state to the upstream request stream info, making it possible to look up upstream connection state from upstream HTTP access logs. 

Risk Level: Low
Testing: UT
Docs Changes: n/a
Release Notes: n/a
Fixes #10457

### CODE REVIEW

File: include/envoy/stream_info/stream_info.h [hunk 1]
```diff
@@ -479,8 +479,7 @@ class StreamInfo {
    * filters (append only). Both object types can be consumed by multiple filters.
    * @return the filter state associated with this request.
    */
-  virtual const FilterStateSharedPtr& filterState() PURE;
-  virtual const FilterState& filterState() const PURE;
+  virtual const FilterStateSharedPtr& filterState() const PURE;
```
Comment (snowp) [inline include/envoy/stream_info/stream_info.h]: This change is a bit unfortunate; it's to make it possible to get the parent `FilterStateSharedPtr` from a const handle. Open for suggestions here for how to do this better

Comment (ggreenway) [reply include/envoy/stream_info/stream_info.h]: I'd rather have a const-cast somewhere; this looks on the surface like it is const-safe, but the returned ptr is non-const, even though the method is const.

Comment (ggreenway) [reply include/envoy/stream_info/stream_info.h]: Actually, I think I would make `onPoolReady()` pass a non-const StreamInfo. I think it's reasonable for that function to have non-const access to the StreamInfo.

File: test/common/router/router_test.cc [hunk 2]
```diff
@@ -3826,6 +3828,58 @@ TEST_F(RouterTest, DirectResponseWithoutLocation) {
...
+// state.
+TEST_F(RouterTest, PropagatesUpstreamFilterState) {
+  NiceMock<Http::MockRequestEncoder> encoder;
+  Http::ResponseDecoder* response_decoder = nullptr;
+
+  // This pattern helps ensure that we're actually invoking the callback.
+  bool filter_state_verified = false;
+  router_.config().upstream_logs_.push_back(
+      std::make_shared<TestAccessLog>([&](const auto& stream_info) {
+        filter_state_verified = stream_info.upstreamFilterState()->hasDataWithName("foo");
```
Comment (ggreenway) [inline test/common/router/router_test.cc]: nit: would slightly increase readability if the string was "upstream data" or similar.

File: source/extensions/access_loggers/grpc/grpc_access_log_utils.cc [hunk 3]
```diff
@@ -239,9 +239,9 @@ void Utility::extractCommonAccessLogProperties(
   }
 
   for (const auto& key : config.filter_state_objects_to_log()) {
-    if (stream_info.filterState().hasDataWithName(key)) {
+    if (stream_info.filterState()->hasDataWithName(key)) {
```
Comment (sunjayBhatia) [inline source/extensions/access_loggers/grpc/grpc_access_log_utils.cc]: It appears https://github.com/envoyproxy/envoy/pull/9845 added this ambiguity causing the windows CI compilation failures, not yet quite sure why MSVC is able to resolve it here: https://github.com/envoyproxy/envoy/pull/9845/files#diff-a5570d8b9053a47a1547728cb6c90989 and not this case

**Comment by ggreenway** [comment]
CI is unhappy :(

/wait

**Comment by snowp** [comment]
Seems to be OOMing in CI, merging again to pick up https://github.com/envoyproxy/envoy/pull/10669

**Comment by snowp** [comment]
@ggreenway I think this can be reviewed now, the failing PR run seems to be due to general OS X issues in CI

**Pull request merged**